File: lib/installer.mjs
import { access, copyFile, cp, mkdir, readFile, readdir, rename, rm, stat } from "node:fs/promises";
import os from "node:os";
import path from "node:path";

import { discoverExtensions } from "./extensions.mjs";
import { INTERNAL_MANIFEST, verifySkillManifest } from "./manifest.mjs";

export const SKILL_NAME = "creating-explainer-videos";
export const LEGACY_SKILL_NAME = "creating-ai-principle-videos";

async function exists(filePath) {
  try {
    await access(filePath);
    return true;
  } catch {
    return false;
  }
}

export function resolveCodexHome(explicitHome) {
  return path.resolve(explicitHome || process.env.CODEX_HOME || path.join(os.homedir(), ".codex"));
}

export function resolveSkillsRoot(options = {}) {
  if (typeof options === "string") {
    return path.join(resolveCodexHome(options), "skills");
  }
  if (options.destination) {
    return path.resolve(options.destination);
  }
  const target = options.target || "codex";
  if (target !== "codex") {
    throw new Error(`target ${target} requires --destination`);
  }
  return path.join(resolveCodexHome(options.codexHome), "skills");
}

export function installedSkillPath(options = {}) {
  return path.join(resolveSkillsRoot(options), SKILL_NAME);
}

function backupRoot(options, skillsRoot) {
  if (typeof options !== "string" && options.destination) {
    return path.join(path.dirname(skillsRoot), ".explainer-video-skill-backups");
  }
  return path.join(resolveCodexHome(typeof options === "string" ? options : options.codexHome), "skill-backups");
}

function timestamp() {
  return new Date().toISOString().replace(/[:.]/g, "-");
}

async function validateSkillRoot(skillRoot) {
  const root = path.resolve(skillRoot);
  const errors = [];
  let skillText = "";
  try {
    skillText = await readFile(path.join(root, "SKILL.md"), "utf8");
  } catch (error) {
    errors.push(`cannot read SKILL.md: ${error.message}`);
  }
  if (!/^---\s*[\s\S]*?\bname:\s*creating-explainer-videos\s*[\s\S]*?---/m.test(skillText)) {
    errors.push("SKILL.md frontmatter does not declare creating-explainer-videos");
  }
  for (const directory of ["agents", "scripts", "references"]) {
    if (!(await exists(path.join(root, directory)))) {
      errors.push(`missing required directory: ${directory}`);
    }
  }
  const extensions = await discoverExtensions(root);
  for (const extension of extensions) {
    if (!extension.valid) {
      errors.push(`invalid extension ${extension.id || path.basename(extension.root)}: ${extension.errors.join("; ")}`);
    }
  }
  return { valid: errors.length === 0, root, errors, extensions };
}

export async function verifyInstalledSkill(options = {}) {
  const target = installedSkillPath(options);
  if (!(await exists(target))) {
    return { valid: false, target, errors: ["skill is not installed"], extensions: [] };
  }
  const validation = await validateSkillRoot(target);
  const manifestPath = path.join(target, INTERNAL_MANIFEST);
  const integrity = await exists(manifestPath)
    ? await verifySkillManifest({ skillRoot: target, manifestPath })
    : { valid: true, checkedFiles: 0, errors: [], skipped: true };
  return {
    ...validation,
    valid: validation.valid && integrity.valid,
    errors: [...validation.errors, ...integrity.errors],
    integrity,
    target,
  };
}

export async function installSkill({ source, codexHome, destination, target: targetKind, manifestPath } = {}) {
  if (!source) throw new Error("source is required");
  const sourceRoot = path.resolve(source);
  const validation = await validateSkillRoot(sourceRoot);
  if (!validation.valid) {
    throw new Error(`Invalid skill source: ${validation.errors.join("; ")}`);
  }
  if (manifestPath) {
    const integrity = await verifySkillManifest({ skillRoot: sourceRoot, manifestPath });
    if (!integrity.valid) {
      throw new Error(`Skill package integrity check failed: ${integrity.errors.join("; ")}`);
    }
  }

  const installOptions = { codexHome, destination, target: targetKind };
  const skillsRoot = resolveSkillsRoot(installOptions);
  const backupsRoot = backupRoot(installOptions, skillsRoot);
  const target = path.join(skillsRoot, SKILL_NAME);
  const legacyTarget = path.join(skillsRoot, LEGACY_SKILL_NAME);
  if (path.resolve(sourceRoot) === path.resolve(target)) {
    throw new Error("source and installed target are the same directory");
  }
  await mkdir(skillsRoot, { recursive: true });
  await mkdir(backupsRoot, { recursive: true });

  const hadExisting = await exists(target);
  const hadLegacy = !hadExisting && await exists(legacyTarget);
  const backupPath = hadExisting
    ? path.join(backupsRoot, `${SKILL_NAME}-${timestamp()}-${process.pid}`)
    : null;
  const legacyBackupPath = hadLegacy
    ? path.join(backupsRoot, `${LEGACY_SKILL_NAME}-migration-${timestamp()}-${process.pid}`)
    : null;
  const staging = path.join(skillsRoot, `.${SKILL_NAME}-install-${process.pid}-${Date.now()}`);

  try {
    await rm(staging, { recursive: true, force: true });
    await cp(sourceRoot, staging, { recursive: true, force: false, errorOnExist: true });
    if (manifestPath) {
      await copyFile(path.resolve(manifestPath), path.join(staging, INTERNAL_MANIFEST));
    }
    const stagedValidation = await validateSkillRoot(staging);
    if (!stagedValidation.valid) {
      throw new Error(`staged skill failed validation: ${stagedValidation.errors.join("; ")}`);
    }
    if (hadExisting) {
      await rename(target, backupPath);
    }
    if (hadLegacy) {
      await rename(legacyTarget, legacyBackupPath);
    }
    await rename(staging, target);
    return {
      action: hadExisting ? "updated" : hadLegacy ? "migrated" : "installed",
      targetKind: destination ? "custom" : "codex",
      skillsRoot,
      target,
      backupPath,
      legacyBackupPath,
      extensions: stagedValidation.extensions.map(({ id, type, version }) => ({ id, type, version })),
    };
  } catch (error) {
    await rm(staging, { recursive: true, force: true });
    if (backupPath && (await exists(backupPath)) && !(await exists(target))) {
      await rename(backupPath, target);
    }
    if (legacyBackupPath && (await exists(legacyBackupPath)) && !(await exists(legacyTarget))) {
      await rename(legacyBackupPath, legacyTarget);
    }
    throw error;
  }
}

export async function uninstallSkill(options = {}) {
  const skillsRoot = resolveSkillsRoot(options);
  const target = installedSkillPath(options);
  if (!(await exists(target))) {
    throw new Error(`skill is not installed: ${target}`);
  }
  const backupsRoot = backupRoot(options, skillsRoot);
  await mkdir(backupsRoot, { recursive: true });
  const backupPath = path.join(backupsRoot, `${SKILL_NAME}-uninstalled-${timestamp()}-${process.pid}`);
  await rename(target, backupPath);
  return { action: "uninstalled", targetKind: options.destination ? "custom" : "codex", skillsRoot, target, backupPath };
}

export async function rollbackSkill(options = {}) {
  const skillsRoot = resolveSkillsRoot(options);
  const target = installedSkillPath(options);
  const backupsRoot = backupRoot(options, skillsRoot);
  if (!(await exists(backupsRoot))) {
    throw new Error(`no backup directory exists: ${backupsRoot}`);
  }

  let selected;
  if (options.backupPath) {
    selected = path.resolve(options.backupPath);
    if (path.dirname(selected) !== path.resolve(backupsRoot) || !path.basename(selected).startsWith(`${SKILL_NAME}-`)) {
      throw new Error("backup must be an exact creating-explainer-videos backup in this target's backup directory");
    }
  } else {
    const candidates = [];
    for (const entry of await readdir(backupsRoot, { withFileTypes: true })) {
      if (!entry.isDirectory() || !entry.name.startsWith(`${SKILL_NAME}-`)) continue;
      const candidate = path.join(backupsRoot, entry.name);
      candidates.push({ path: candidate, modified: (await stat(candidate)).mtimeMs });
    }
    candidates.sort((left, right) => right.modified - left.modified || right.path.localeCompare(left.path));
    selected = candidates[0]?.path;
  }
  if (!selected || !(await exists(selected))) {
    throw new Error(`no recoverable backup found in: ${backupsRoot}`);
  }
  const validation = await validateSkillRoot(selected);
  if (!validation.valid) {
    throw new Error(`backup is not a valid skill: ${validation.errors.join("; ")}`);
  }

  await mkdir(skillsRoot, { recursive: true });
  const hadExisting = await exists(target);
  const displacedTo = hadExisting
    ? path.join(backupsRoot, `${SKILL_NAME}-before-rollback-${timestamp()}-${process.pid}`)
    : null;
  try {
    if (hadExisting) await rename(target, displacedTo);
    await rename(selected, target);
  } catch (error) {
    if (displacedTo && (await exists(displacedTo)) && !(await exists(target))) {
      await rename(displacedTo, target);
    }
    throw error;
  }
  return {
    action: "rolled_back",
    targetKind: options.destination ? "custom" : "codex",
    skillsRoot,
    target,
    restoredFrom: selected,
    displacedTo,
  };
}

File: python_src/creating_explainer_videos_skill/installer.py
from __future__ import annotations

from datetime import datetime, timezone
import json
import os
from pathlib import Path
import re
import shutil
from typing import Any

from .integrity import INTERNAL_MANIFEST, list_skill_files, verify_skill_manifest


SKILL_NAME = "creating-explainer-videos"
LEGACY_SKILL_NAME = "creating-ai-principle-videos"
EXTENSION_API_VERSION = 1
EXTENSION_TYPES = {"visual", "voice", "research", "qc", "publishing"}


def resolve_codex_home(explicit_home: str | Path | None = None) -> Path:
    return Path(explicit_home or os.environ.get("CODEX_HOME") or Path.home() / ".codex").resolve()


def resolve_skills_root(
    *,
    codex_home: str | Path | None = None,
    destination: str | Path | None = None,
    target: str | None = None,
) -> Path:
    if destination is not None:
        return Path(destination).resolve()
    selected_target = target or "codex"
    if selected_target != "codex":
        raise ValueError(f"target {selected_target} requires --destination")
    return resolve_codex_home(codex_home) / "skills"


def installed_skill_path(**options: Any) -> Path:
    return resolve_skills_root(**options) / SKILL_NAME


def _backup_root(*, skills_root: Path, codex_home: str | Path | None, destination: str | Path | None) -> Path:
    if destination is not None:
        return skills_root.parent / ".explainer-video-skill-backups"
    return resolve_codex_home(codex_home) / "skill-backups"


def _timestamp() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H-%M-%S-%fZ")


def _resolve_entrypoint(extension_root: Path, relative: str) -> tuple[Path, bool]:
    if not relative or Path(relative).is_absolute():
        return extension_root, False
    resolved = (extension_root / relative).resolve()
    try:
        resolved.relative_to(extension_root.resolve())
    except ValueError:
        return resolved, False
    return resolved, resolved != extension_root.resolve()


def validate_extension(extension_root: Path) -> dict[str, Any]:
    root = extension_root.resolve()
    errors: list[str] = []
    manifest_path = root / "extension.json"
    try:
        manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        return {
            "valid": False,
            "root": str(root),
            "manifestPath": str(manifest_path),
            "errors": [f"cannot read extension.json: {error}"],
        }

    extension_id = manifest.get("id")
    extension_type = manifest.get("type")
    if manifest.get("apiVersion") != EXTENSION_API_VERSION:
        errors.append(
            f"unsupported apiVersion {manifest.get('apiVersion')}; expected {EXTENSION_API_VERSION}"
        )
    if not isinstance(extension_id, str) or not re.fullmatch(r"[a-z0-9][a-z0-9-]*", extension_id):
        errors.append("id must use lowercase letters, digits, and hyphens")
    if extension_type not in EXTENSION_TYPES:
        errors.append(f"unsupported extension type: {extension_type}")
    if not isinstance(manifest.get("version"), str) or not re.fullmatch(r"\d+\.\d+\.\d+", manifest["version"]):
        errors.append("version must use MAJOR.MINOR.PATCH")
    for field in ("displayName", "description"):
        if not isinstance(manifest.get(field), str) or not manifest[field].strip():
            errors.append(f"{field} must be a non-empty string")
    for field in ("capabilities", "permissions"):
        if not isinstance(manifest.get(field), list) or not all(isinstance(item, str) for item in manifest[field]):
            errors.append(f"{field} must be an array of strings")

    entrypoints = manifest.get("entrypoints")
    if not isinstance(entrypoints, dict):
        errors.append("entrypoints must be an object")
    else:
        declared: list[str] = []
        for field in ("reference", "profile"):
            value = entrypoints.get(field)
            if not isinstance(value, str):
                errors.append(f"entrypoints.{field} must be a string")
            else:
                declared.append(value)
        assets = entrypoints.get("assets", [])
        if not isinstance(assets, list) or not all(isinstance(item, str) for item in assets):
            errors.append("entrypoints.assets must be an array of strings")
        else:
            declared.extend(assets)
        for relative in declared:
            resolved, inside = _resolve_entrypoint(root, relative)
            if not inside:
                errors.append(f"entrypoint is outside extension directory: {relative}")
            elif not resolved.is_file():
                errors.append(f"missing entrypoint: {relative}")

    return {
        "id": extension_id,
        "type": extension_type,
        "version": manifest.get("version"),
        "displayName": manifest.get("displayName"),
        "permissions": manifest.get("permissions", []),
        "valid": not errors,
        "errors": errors,
        "root": str(root),
        "manifestPath": str(manifest_path),
    }


def discover_extensions(skill_root: str | Path) -> list[dict[str, Any]]:
    extensions_root = Path(skill_root).resolve() / "extensions"
    if not extensions_root.is_dir():
        return []
    return [
        validate_extension(path)
        for path in sorted(extensions_root.iterdir(), key=lambda item: item.name)
        if path.is_dir()
    ]


def validate_skill_root(skill_root: str | Path) -> dict[str, Any]:
    root = Path(skill_root).resolve()
    errors: list[str] = []
    try:
        skill_text = (root / "SKILL.md").read_text(encoding="utf-8")
    except OSError as error:
        skill_text = ""
        errors.append(f"cannot read SKILL.md: {error}")
    if not re.search(
        r"^---\s*[\s\S]*?\bname:\s*creating-explainer-videos\s*[\s\S]*?---",
        skill_text,
        re.MULTILINE,
    ):
        errors.append("SKILL.md frontmatter does not declare creating-explainer-videos")
    for directory in ("agents", "scripts", "references"):
        if not (root / directory).is_dir():
            errors.append(f"missing required directory: {directory}")
    try:
        list_skill_files(root)
    except (OSError, ValueError) as error:
        errors.append(str(error))
    extensions = discover_extensions(root)
    for extension in extensions:
        if not extension["valid"]:
            errors.append(
                f"invalid extension {extension.get('id') or Path(extension['root']).name}: "
                + "; ".join(extension["errors"])
            )
    return {"valid": not errors, "root": str(root), "errors": errors, "extensions": extensions}


def verify_installed_skill(
    *,
    codex_home: str | Path | None = None,
    destination: str | Path | None = None,
    target: str | None = None,
) -> dict[str, Any]:
    install_options = {"codex_home": codex_home, "destination": destination, "target": target}
    installed = installed_skill_path(**install_options)
    if not installed.is_dir():
        return {"valid": False, "target": str(installed), "errors": ["skill is not installed"], "extensions": []}
    validation = validate_skill_root(installed)
    internal_manifest = installed / INTERNAL_MANIFEST
    integrity = (
        verify_skill_manifest(installed, internal_manifest)
        if internal_manifest.is_file()
        else {"valid": True, "checkedFiles": 0, "errors": [], "skipped": True}
    )
    errors = [*validation["errors"], *integrity["errors"]]
    return {
        **validation,
        "valid": validation["valid"] and integrity["valid"],
        "errors": errors,
        "integrity": integrity,
        "target": str(installed),
    }


def install_skill(
    source: str | Path,
    *,
    codex_home: str | Path | None = None,
    destination: str | Path | None = None,
    target: str | None = None,
    manifest_path: str | Path | None = None,
) -> dict[str, Any]:
    source_root = Path(source).resolve()
    validation = validate_skill_root(source_root)
    if not validation["valid"]:
        raise ValueError("Invalid skill source: " + "; ".join(validation["errors"]))
    if manifest_path is not None:
        integrity = verify_skill_manifest(source_root, Path(manifest_path).resolve())
        if not integrity["valid"]:
            raise ValueError("Skill package integrity check failed: " + "; ".join(integrity["errors"]))

    install_options = {"codex_home": codex_home, "destination": destination, "target": target}
    skills_root = resolve_skills_root(**install_options)
    installed = skills_root / SKILL_NAME
    legacy = skills_root / LEGACY_SKILL_NAME
    if source_root == installed.resolve():
        raise ValueError("source and installed target are the same directory")
    backups = _backup_root(skills_root=skills_root, codex_home=codex_home, destination=destination)
    skills_root.mkdir(parents=True, exist_ok=True)
    backups.mkdir(parents=True, exist_ok=True)

    had_existing = installed.is_dir()
    had_legacy = not had_existing and legacy.is_dir()
    unique = f"{_timestamp()}-{os.getpid()}"
    backup = backups / f"{SKILL_NAME}-{unique}" if had_existing else None
    legacy_backup = backups / f"{LEGACY_SKILL_NAME}-migration-{unique}" if had_legacy else None
    staging = skills_root / f".{SKILL_NAME}-install-{unique}"

    try:
        if staging.exists():
            shutil.rmtree(staging)
        shutil.copytree(source_root, staging)
        if manifest_path is not None:
            shutil.copy2(Path(manifest_path).resolve(), staging / INTERNAL_MANIFEST)
        staged = validate_skill_root(staging)
        if not staged["valid"]:
            raise ValueError("staged skill failed validation: " + "; ".join(staged["errors"]))
        if manifest_path is not None:
            staged_integrity = verify_skill_manifest(staging, staging / INTERNAL_MANIFEST)
            if not staged_integrity["valid"]:
                raise ValueError("staged skill failed integrity verification: " + "; ".join(staged_integrity["errors"]))
        if backup is not None:
            installed.rename(backup)
        if legacy_backup is not None:
            legacy.rename(legacy_backup)
        staging.rename(installed)
    except Exception:
        if staging.exists():
            shutil.rmtree(staging)
        if backup is not None and backup.exists() and not installed.exists():
            backup.rename(installed)
        if legacy_backup is not None and legacy_backup.exists() and not legacy.exists():
            legacy_backup.rename(legacy)
        raise

    return {
        "action": "updated" if had_existing else "migrated" if had_legacy else "installed",
        "targetKind": "custom" if destination is not None else "codex",
        "skillsRoot": str(skills_root),
        "target": str(installed),
        "backupPath": str(backup) if backup is not None else None,
        "legacyBackupPath": str(legacy_backup) if legacy_backup is not None else None,
        "extensions": [
            {"id": item["id"], "type": item["type"], "version": item["version"]}
            for item in staged["extensions"]
        ],
    }


def uninstall_skill(
    *,
    codex_home: str | Path | None = None,
    destination: str | Path | None = None,
    target: str | None = None,
) -> dict[str, Any]:
    skills_root = resolve_skills_root(codex_home=codex_home, destination=destination, target=target)
    installed = skills_root / SKILL_NAME
    if not installed.is_dir():
        raise ValueError(f"skill is not installed: {installed}")
    backups = _backup_root(skills_root=skills_root, codex_home=codex_home, destination=destination)
    backups.mkdir(parents=True, exist_ok=True)
    backup = backups / f"{SKILL_NAME}-uninstalled-{_timestamp()}-{os.getpid()}"
    installed.rename(backup)
    return {
        "action": "uninstalled",
        "targetKind": "custom" if destination is not None else "codex",
        "skillsRoot": str(skills_root),
        "target": str(installed),
        "backupPath": str(backup),
    }


def rollback_skill(
    *,
    codex_home: str | Path | None = None,
    destination: str | Path | None = None,
    target: str | None = None,
    backup_path: str | Path | None = None,
) -> dict[str, Any]:
    skills_root = resolve_skills_root(codex_home=codex_home, destination=destination, target=target)
    installed = skills_root / SKILL_NAME
    backups = _backup_root(skills_root=skills_root, codex_home=codex_home, destination=destination)
    if not backups.is_dir():
        raise ValueError(f"no backup directory exists: {backups}")

    if backup_path is not None:
        selected = Path(backup_path).resolve()
        if selected.parent != backups.resolve() or not selected.name.startswith(f"{SKILL_NAME}-"):
            raise ValueError("backup must be an exact creating-explainer-videos backup in this target's backup directory")
    else:
        candidates = [
            path for path in backups.iterdir()
            if path.is_dir() and path.name.startswith(f"{SKILL_NAME}-")
        ]
        candidates.sort(key=lambda path: (path.stat().st_mtime_ns, path.name), reverse=True)
        selected = candidates[0] if candidates else None
    if selected is None or not selected.is_dir():
        raise ValueError(f"no recoverable backup found in: {backups}")
    validation = validate_skill_root(selected)
    if not validation["valid"]:
        raise ValueError("backup is not a valid skill: " + "; ".join(validation["errors"]))
    internal_manifest = selected / INTERNAL_MANIFEST
    if internal_manifest.is_file():
        integrity = verify_skill_manifest(selected, internal_manifest)
        if not integrity["valid"]:
            raise ValueError("backup failed integrity verification: " + "; ".join(integrity["errors"]))

    skills_root.mkdir(parents=True, exist_ok=True)
    displaced = backups / f"{SKILL_NAME}-before-rollback-{_timestamp()}-{os.getpid()}" if installed.is_dir() else None
    try:
        if displaced is not None:
            installed.rename(displaced)
        selected.rename(installed)
    except Exception:
        if displaced is not None and displaced.exists() and not installed.exists():
            displaced.rename(installed)
        raise
    return {
        "action": "rolled_back",
        "targetKind": "custom" if destination is not None else "codex",
        "skillsRoot": str(skills_root),
        "target": str(installed),
        "restoredFrom": str(selected),
        "displacedTo": str(displaced) if displaced is not None else None,
    }
